teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+ 
+2. Interceptors are configured or installed via an Interceptor CRD that describes an address for the Interceptor as well as  a list of valid parameters that the interceptor accepts.
+
+3. Built-in interceptors are still packaged as part of Triggers but run as separate processes instead of in the controller.
+
+4. Users configure interceptors in their Triggers and EventListeners. There will be a change in the syntax but we will support both the old and new syntax to allow for backwards compatibility:
+
+```yaml
+# OLD:
+interceptors:
+  - cel:
+    filter: "header.match("EventType", "push")
+
+# NEW:
+interceptors:
+  - type: CEL
+    params:
+    - name: "filter"
+      value: header.match("EventType", "push") 
+```
+
+
+5. For backwards compatibility, we will keep supporting the old WebhookInterceptor interface until beta. However, we will recommend users use the new Interceptor interface instead for writing any new Interceptors.
+
+See the [Design Details](#design-details) section below for more.
+
+
+### Notes/Constraints/Caveats
+
+1. Deployment Footprint and Scaling Concerns
+
+  The proposal suggests one interceptor deployment per cluster. This can cause scaling concerns. With one interceptor serving multiple EventListeners, it is possible that one EL that has many incoming events (or has many triggers with many interceptors) starts making too many requests to the interceptor. This could result in Listeners being bottle-necked on the interceptor processing and then failing to process their triggers (or building up a large queue of triggers/events to process).
+  
+  The proposal also increases the deployment footprint for Triggers since now the 4 built-in interceptors are deployed as separate pods. In practice, these should be fairly small stateless services that are only created once per cluster. In addition, a HorizontalPodAutoscaler can be configured to scale the resources if needed. Finally, we could also consider bundling all 4 built-in interceptors into one service. 
+
+2. Logging/Debugging
+
+    Fetching logs from the Interceptor pods to debug a failed trigger is going to be cumbersome. So, each interceptor should return a descriptive errorMessage/status/summary in its response that will be logged by the EventListener as well as sent back as a response to the incoming event
```

Comment:
in the long run it probably makes sense to have some kind of ID (EventID?) that gets passed through to each interceptor and can be used to piece together a complete picture of what happened for a particular event.

it might make sense to one day investigate (probably not as part of this TEP!) what we'd need to require from interceptors to support integrating them with a distributed tracing system (e.g. https://zipkin.io/)

Reply:
we are already passing the EventID as part of each request.  Are you suggesting a per Interceptor id of some sort?


(AdmissionWebhooks also have some interesting features for metircs/tracing: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#monitoring-admission-webhooks)

Reply:
I don't know enough about it - I'd guess that EventID would do the trick?

Anyway this is definitely a segue/scope creep XD

Reply:
yeah, I expect there will be a few follow up TEPs (like this) after this.